docker_util.py

```diff
@@ -0,0 +1,84 @@
+import json
+import subprocess
+
+from verbose_print import print_v
+
+
+class ImageNotFoundError(Exception):
+    def __init__(self, image):
+        self.image = image
+        self.message = f'The image {image} was not found.'
+
+    def __str__(self):
+        return self.message
+
+
+def tag_image(source, target, is_dry_run):
+    print(f'Tagging image {source} as {target}')
+    tag_command = ('docker', 'tag', source, target)
+    if is_dry_run:
+        print_v(' '.join(tag_command))
+        print('Image was not actually tagged since this is a dry run')
+    else:
+        _check_output_and_print(tag_command)
+
+
+def pull_image(image):
+    pull_command = ('docker', 'pull', image)
+    try:
+        _check_output_and_print(pull_command)
+    except subprocess.CalledProcessError as e:
+        stderr = e.stderr.decode()
+        if image in stderr and 'not found' in stderr:
+            raise ImageNotFoundError(image) from e
+        else:
+            raise
+
+
+def push_image(image, is_dry_run):
+    print(f'Pushing image {image} ...')
+    push_command = ('docker', 'push', image)
+    if is_dry_run:
+        print_v(' '.join(push_command))
+        print('Image was not actually pushed since this is a dry run')
+    else:
+        _check_output_and_print(push_command)
+
+
+def inspect_image(image):
+    inspect_command = ('docker', 'inspect', image)
+    try:
+        docker_inspect_output = _check_output_and_print(inspect_command)
+    except subprocess.CalledProcessError as e:
+        if 'Error: No such object' in e.stderr.decode():
+            raise ImageNotFoundError(image) from e
+        else:
+            raise
+    return json.loads(docker_inspect_output)[0]
```

Comment:
leaving a note to myself to come back to this if inspect isn't actually necessary (I don't think it is?)

Reply:
I mostly only used inspect to verify an image exists, and as a circuitous way to validate the source image param. It's also used here and there during testing, but I can just move it over to testing code.

I'm not sure what the harm is in using `docker inspect` though? It's probably the simplest way to verify an image exists.